File: packages/swift/Sources/EcollectSDK/Types/PaymentIntent.swift
import Foundation

/// Represents a payment intention to be processed by ecollect.
public struct PaymentIntent {
    /// Total amount without taxes.
    public var amount: Decimal
    /// VAT/tax amount (optional).
    public var vatAmount: Decimal?
    /// ISO 4217 currency code (e.g. "COP", "USD", "MXN", "DOP").
    public var currency: String
    /// Optional ecollect service code override (falls back to client config).
    public var srvCode: Int?
    /// Redirect URL after hosted payment completion.
    public var urlRedirect: String?
    /// Webhook URL for transaction result notification.
    public var urlResponse: String?
    /// ISO 639-1 language code ("ES" or "EN"). Defaults to "ES".
    public var langCode: String
    /// ecollect payment system code (see PaymentSystem Values).
    public var paymentSystem: String?
    /// Financial institution code selected by the user.
    public var fiCode: String?
    /// Invoice reference for in-person payment channels.
    public var invoice: String?
    /// Invoice due date in yyyyMMddHHmmss format.
    public var invoiceDueDate: String?
    /// Policy code for debit/credit policy configuration.
    public var policyCode: String?
    /// Request type: 0=immediate, 1=pre-auth, TicketId=capture, -TicketId=void.
    public var requestType: Int
    /// Reference array: [0]=DocType, [1]=DocNumber, [2]=MerchantOrderId,
    /// [3]=FullName, [4]=Email, [5]=Phone, [6..n]=extra.
    public var referenceArray: [String]
    /// Additional payment attributes (IPAddress, DeviceFingerPrint, etc.).
    public var paymentInfoArray: [PaymentInfoType]?
    /// Token info for credit card token-based payments.
    public var tokenInfoArray: [PaymentInfoType]?
    /// Subservices for payment splitting/dispersal.
    public var subservicesArray: [SubserviceType]?
    /// Channel-specific info.
    public var channelInfoArray: [ChannelInfoType]?

    public init(
        amount: Decimal,
        currency: String,
        referenceArray: [String],
        vatAmount: Decimal? = nil,
        srvCode: Int? = nil,
        urlRedirect: String? = nil,
        urlResponse: String? = nil,
        langCode: String = "ES",
        paymentSystem: String? = nil,
        fiCode: String? = nil,
        invoice: String? = nil,
        invoiceDueDate: String? = nil,
        policyCode: String? = nil,
        requestType: Int = 0,
        paymentInfoArray: [PaymentInfoType]? = nil,
        tokenInfoArray: [PaymentInfoType]? = nil,
        subservicesArray: [SubserviceType]? = nil,
        channelInfoArray: [ChannelInfoType]? = nil
    ) {
        self.amount = amount
        self.currency = currency
        self.referenceArray = referenceArray
        self.vatAmount = vatAmount
        self.srvCode = srvCode
        self.urlRedirect = urlRedirect
        self.urlResponse = urlResponse
        self.langCode = langCode
        self.paymentSystem = paymentSystem
        self.fiCode = fiCode
        self.invoice = invoice
        self.invoiceDueDate = invoiceDueDate
        self.policyCode = policyCode
        self.requestType = requestType
        self.paymentInfoArray = paymentInfoArray
        self.tokenInfoArray = tokenInfoArray
        self.subservicesArray = subservicesArray
        self.channelInfoArray = channelInfoArray
    }
}

File: packages/php/src/Types/PaymentIntent.php
<?php

declare(strict_types=1);

namespace Ecollect\Types;

/**
 * Represents a payment intention with all transaction parameters.
 */
class PaymentIntent
{
    /** @var float */
    public $amount;

    /** @var string ISO 4217 currency code */
    public $currency;

    /** @var Customer */
    public $customer;

    /** @var int|null */
    public $srvCode;

    /** @var string|null */
    public $merchantTransactionId;

    /** @var string|null */
    public $paymentSystem;

    /** @var string|null */
    public $fiCode;

    /** @var string|null */
    public $tokenId;

    /** @var string|null */
    public $secureCode;

    /** @var int|null */
    public $installments;

    /** @var float|null */
    public $vatAmount;

    /** @var string|null */
    public $redirectUrl;

    /** @var string|null */
    public $responseUrl;

    /** @var string|null */
    public $langCode;

    /** @var string|null */
    public $invoice;

    /** @var string|null yyyyMMddHHmmss format */
    public $invoiceDueDate;

    /** @var string|null */
    public $policyCode;

    /** @var string|null */
    public $ipAddress;

    /** @var string|null */
    public $deviceFingerPrint;

    /** @var string|null */
    public $oneTimePassword;

    /** @var string|null '0' or '1' for PSE (Colombia) */
    public $userType;

    /** @var string[]|null */
    public $additionalReferences;

    /** @var int|null Lifetime in seconds for QR payment links */
    public $qrLifetimeSecs;

    /** @var array<int, array{entityCode: int, srvCode: int, valueType: int, transValue: float, transVatValue?: float}>|null */
    public $subservices;

    /** @var array<string, string>|null Extra payment info key=>value pairs */
    public $paymentInfo;

    /**
     * @param array<string, mixed> $data
     */
    public function __construct(array $data = [])
    {
        $this->amount                = (float)($data['amount']                  ?? 0);
        $this->currency              = $data['currency']                         ?? '';
        $this->customer              = $data['customer'] instanceof Customer
            ? $data['customer']
            : new Customer($data['customer'] ?? []);
        $this->srvCode               = isset($data['srv_code'])               ? (int)$data['srv_code']               : null;
        $this->merchantTransactionId = $data['merchant_transaction_id']        ?? null;
        $this->paymentSystem         = isset($data['payment_system'])          ? (string)$data['payment_system']     : null;
        $this->fiCode                = $data['fi_code']                         ?? null;
        $this->tokenId               = $data['token_id']                        ?? null;
        $this->secureCode            = $data['secure_code']                     ?? null;
        $this->installments          = isset($data['installments'])             ? (int)$data['installments']          : null;
        $this->vatAmount             = isset($data['vat_amount'])               ? (float)$data['vat_amount']          : null;
        $this->redirectUrl           = $data['redirect_url']                    ?? null;
        $this->responseUrl           = $data['response_url']                    ?? null;
        $this->langCode              = $data['lang_code']                       ?? null;
        $this->invoice               = $data['invoice']                         ?? null;
        $this->invoiceDueDate        = $data['invoice_due_date']                ?? null;
        $this->policyCode            = $data['policy_code']                     ?? null;
        $this->ipAddress             = $data['ip_address']                      ?? null;
        $this->deviceFingerPrint     = $data['device_finger_print']             ?? null;
        $this->oneTimePassword       = $data['one_time_password']               ?? null;
        $this->userType              = isset($data['user_type'])                ? (string)$data['user_type']          : null;
        $this->additionalReferences  = $data['additional_references']           ?? null;
        $this->qrLifetimeSecs        = isset($data['qr_lifetime_secs'])        ? (int)$data['qr_lifetime_secs']     : null;
        $this->subservices           = $data['subservices']                     ?? null;
        $this->paymentInfo           = $data['payment_info']                    ?? null;
    }
}
